File: 202-happy-number/202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        seen = {n}
        while n > 1:
            total = 0
            while n > 0:
                n, d = divmod(n, 10)
                total += d * d
                
            if total in seen:
                return False
            seen.add(total)
            
            n = total
        return True

File: 202-happy-number/202-happy-number.java
class Solution {
    public boolean isHappy(int n) {
        Set<Integer> seen = new HashSet<>();
        while(n > 1){
            if(seen.contains(n))
                return false;
            seen.add(n);
            
            int total = 0;
            while(n > 0){
                int d = n % 10;
                total += d * d;
                n /= 10;
            }
            n = total;
        }
        return true;
    }
}